File: http/R/httpclient.R
# Copyright Core DF — Apache License 2.0
http_missing_env <- function(vars) list(status_code=601L, error=paste("Environment variables", vars, "should be defined"))
http_transport_error <- function(message="inaccessible") list(status_code=0L, error=message)

http_parse_body <- function(raw) {
  if (is.null(raw) || raw == "") return(NULL)
  out <- tryCatch(jsonlite::fromJSON(raw, simplifyVector=FALSE), error=function(e) raw)
  out
}

http_request <- function(method, url, headers=list(), body=NULL, params=NULL) {
  if (!requireNamespace("httr", quietly=TRUE)) stop("install.packages('httr')")
  if (!requireNamespace("jsonlite", quietly=TRUE)) stop("install.packages('jsonlite')")
  if (length(params)) {
    sep <- if (grepl("\\?", url, fixed=TRUE)) "&" else "?"
    url <- paste0(url, sep, paste(paste0(names(params),"=", params), collapse="&"))
  }
  hdrs <- httr::add_headers(.headers=unlist(headers))
  resp <- tryCatch({
    if (method == "GET") httr::GET(url, hdrs)
    else if (method == "DELETE") httr::DELETE(url, hdrs)
    else if (method == "PUT") httr::PUT(url, hdrs, body=body, encode="raw", httr::content_type("application/json"))
    else httr::POST(url, hdrs, body=body, encode="raw", httr::content_type("application/json"))
  }, error=function(e) NULL)
  if (is.null(resp)) return(http_transport_error())
  code <- as.integer(httr::status_code(resp))
  text <- httr::content(resp, as="text", encoding="UTF-8")
  parsed <- http_parse_body(text)
  if (code >= 400L) return(list(status_code=code, error=if (is.null(parsed)) "inaccessible" else parsed))
  list(status_code=code, body=parsed)
}

Get <- function(url, headers=NULL, params=NULL) http_request("GET", url, headers %||% list(), params=params)
Post <- function(url, json_body=NULL, data=NULL, headers=NULL) {
  headers <- headers %||% list(); body <- NULL
  if (!is.null(json_body)) { headers[["Content-Type"]] <- headers[["Content-Type"]] %||% "application/json"; body <- jsonlite::toJSON(json_body, auto_unbox=TRUE) }
  else if (!is.null(data)) body <- data
  http_request("POST", url, headers, body)
}
Put <- function(url, json_body=NULL, headers=NULL) {
  headers <- headers %||% list(); body <- NULL
  if (!is.null(json_body)) { headers[["Content-Type"]] <- headers[["Content-Type"]] %||% "application/json"; body <- jsonlite::toJSON(json_body, auto_unbox=TRUE) }
  http_request("PUT", url, headers, body)
}
Delete <- function(url, headers=NULL) http_request("DELETE", url, headers %||% list())
`%||%` <- function(a,b) if (is.null(a)) b else a

File: http/Lua/lib/httpclient.lua
-- Copyright Core DF — Apache License 2.0
local result = require("result")
local json = require("json")
local http = {}

local function shell_quote(s)
  return "'" .. s:gsub("'", "'\\''") .. "'"
end

local function parse_body(raw)
  if not raw or raw == "" then return nil end
  local ok, parsed = pcall(json.decode, raw)
  if ok then return parsed end
  return raw
end

local function request(method, url, headers, body)
  local hdr_args = {}
  for k, v in pairs(headers or {}) do
    hdr_args[#hdr_args + 1] = "-H " .. shell_quote(k .. ": " .. v)
  end
  local body_arg = body and ("-d " .. shell_quote(body)) or ""
  local cmd = string.format("curl -s -S -w '\\n%%{http_code}' -X %s %s %s %s",
    method, table.concat(hdr_args, " "), body_arg, shell_quote(url))
  local fh = io.popen(cmd)
  if not fh then return result.transport_error("curl failed") end
  local raw = fh:read("*a")
  fh:close()
  local code_str = raw:match("(\n%d+)$")
  if not code_str then return result.transport_error("bad curl response") end
  local code = tonumber(code_str:match("(%d+)$")) or 0
  local resp_body = raw:sub(1, #raw - #code_str)
  local parsed = parse_body(resp_body)
  if code >= 400 then
    return { status_code = code, error = parsed or "inaccessible" }
  end
  return { status_code = code, body = parsed }
end

function http.Get(url, headers, params)
  if params then
    local qs = {}
    for k, v in pairs(params) do qs[#qs+1] = k .. "=" .. v end
    url = url .. (url:find("?") and "&" or "?") .. table.concat(qs, "&")
  end
  return request("GET", url, headers)
end

function http.Post(url, json_body, data, headers)
  headers = headers or {}
  local body
  if json_body then
    headers["Content-Type"] = headers["Content-Type"] or "application/json"
    body = json.encode(json_body)
  else body = data end
  return request("POST", url, headers, body)
end

function http.Put(url, json_body, headers)
  headers = headers or {}
  local body
  if json_body then
    headers["Content-Type"] = headers["Content-Type"] or "application/json"
    body = json.encode(json_body)
  end
  return request("PUT", url, headers, body)
end

function http.Delete(url, headers)
  return request("DELETE", url, headers)
end

return http
